File: EMR_BMED.Backend/Controllers/PrescriptionController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

using EMR_BMED.Backend.Models;
using EMR_BMED.Backend.Services;
using EMR_BMED.Backend.Utils;

namespace EMR_BMED.Backend.Controllers
{
  [ApiController]
  [Route("[controller]")]
  public class PrescriptionsController(PrescriptionService prescriptionService) : ControllerBase
  {
    [Authorize]
    [HttpGet("{id}")]
    public async Task<IActionResult> GetOne(
      [FromHeader(Name = "Authorization")] string authHeader,
      [FromRoute] Guid id)
    {
      string token = authHeader.Split(' ').LastOrDefault()!;
      Guid myId = TokenUtils.ExtractId(token);

      try
      {
        PrescriptionModel prescription = await prescriptionService.GetOneAsync(id);
        if (myId != prescription.PatientId && myId != prescription.DoctorId)
        {
          return Forbid();
        }
        return Ok(prescription);
      }
      catch (KeyNotFoundException ex)
      {
        return NotFound(new { message = ex.Message });
      }
    }

    [Authorize]
    [HttpGet("patient/{id}")]
    public async Task<IActionResult> GetAllByPatient(
      [FromHeader(Name = "Authorization")] string authHeader,
      [FromRoute] Guid id)
    {
      string token = authHeader.Split(' ').LastOrDefault()!;
      Guid myId = TokenUtils.ExtractId(token);
      if (myId != id)
      {
        return Forbid();
      }

      try
      {
        PrescriptionModel[] prescriptions = await prescriptionService
          .GetAllByPatientAsync(id);
        return Ok(prescriptions);
      }
      catch (KeyNotFoundException ex)
      {
        return NotFound(new { message = ex.Message });
      }
    }

    [Authorize]
    [HttpGet("doctor/{id}")]
    public async Task<IActionResult> GetAllByDoctor(
      [FromHeader(Name = "Authorization")] string authHeader,
      [FromRoute] Guid id)
    {
      string token = authHeader.Split(' ').LastOrDefault()!;
      Guid myId = TokenUtils.ExtractId(token);
      if (myId != id)
      {
        return Forbid();
      }

      try
      {
        PrescriptionModel[] prescriptions = await prescriptionService
          .GetAllByDoctorAsync(id);
        return Ok(prescriptions);
      }
      catch (KeyNotFoundException ex)
      {
        return NotFound(new { message = ex.Message });
      }
    }

    [Authorize]
    [HttpPost]
    public async Task<IActionResult> Create(
      [FromHeader(Name = "Authorization")] string authHeader,
      [FromBody] PrescriptionCreateDto dto)
    {
      string token = authHeader.Split(' ').LastOrDefault()!;
      Guid myId = TokenUtils.ExtractId(token);
      if (myId != dto.DoctrorId)
      {
        return Forbid();
      }

      try
      {
        await prescriptionService.CreateAsync(dto);
        return Ok();
      }
      catch (KeyNotFoundException ex)
      {
        return NotFound(new { message = ex.Message });
      }
    }

    [Authorize]
    [HttpPut("{id}")]
    public async Task<IActionResult> Update(
      [FromHeader(Name = "Authorization")] string authHeader,
      [FromRoute] Guid id, [FromBody] PrescriptionUpdateDto dto)
    {
      string token = authHeader.Split(' ').LastOrDefault()!;
      Guid myId = TokenUtils.ExtractId(token);
      PrescriptionModel prescription = await prescriptionService.GetOneAsync(id);
      if (myId != prescription.DoctorId)
      {
        return Forbid();
      }

      try
      {
        await prescriptionService.UpdateAsync(id, dto);
        return Ok();
      }
      catch (KeyNotFoundException ex)
      {
        return NotFound(new { message = ex.Message });
      }
    }

    [Authorize]
    [HttpDelete("{id}")]
    public async Task<IActionResult> Delete(
      [FromHeader(Name = "Authorization")] string authHeader,
      [FromRoute] Guid id)
    {
      string token = authHeader.Split(' ').LastOrDefault()!;
      Guid myId = TokenUtils.ExtractId(token);
      PrescriptionModel prescription = await prescriptionService.GetOneAsync(id);
      if (myId != prescription.DoctorId)
      {
        return Forbid();
      }

      try
      {
        await prescriptionService.DeleteAsync(id);
        return Ok();
      }
      catch (KeyNotFoundException ex)
      {
        return NotFound(new { message = ex.Message });
      }
    }
  }
}

File: EMR_BMED.Backend.JS/controllers/prescription.controller.js
import express from "express";
const router = express.Router();

import prescriptionService from "../services/prescription.service.js";
import tokenUtils from "../utils/token.utils.js";

router
  .get("/patient/:id", (req, res) => {
    const id = req.params.id;
    const token = req.headers.authorization.split(" ").at(-1);
    const myId = tokenUtils.extractId(token);
    if (myId !== id) {
      return res.status(403).send({});
    }

    try {
      const prescriptions = prescriptionService.getAllByUser(id, false);
      return res.status(200).send(prescriptions);
    } catch (err) {
      return res.status(404).send({ message: err.message });
    }
  })
  .get("/doctor/:id", (req, res) => {
    const id = req.params.id;
    const token = req.headers.authorization.split(" ").at(-1);
    const myId = tokenUtils.extractId(token);
    if (myId !== id) {
      return res.status(403).send({});
    }

    try {
      const prescriptions = prescriptionService.getAllByUser(id, true);
      return res.status(200).send(prescriptions);
    } catch (err) {
      return res.status(404).send({ message: err.message });
    }
  })
  .get("/:id", (req, res) => {
    const id = req.params.id;
    const token = req.headers.authorization.split(" ").at(-1);
    const myId = tokenUtils.extractId(token);

    try {
      const prescription = prescriptionService.getOne(id);
      if (myId !== prescription.patientId && myId !== prescription.doctorId) {
        return res.status(403).send({});
      }
      return res.status(200).send(prescription);
    } catch (err) {
      return res.status(404).send({ message: err.message });
    }
  })
  .post("/", (req, res) => {
    const token = req.headers.authorization.split(" ").at(-1);
    const myId = tokenUtils.extractId(token);
    if (myId !== req.body.doctorId) {
      return res.status(403).send({});
    }

    try {
      prescriptionService.createOne(req.body);
      return res.status(200).send({});
    } catch (err) {
      return res.status(404).send({ message: err.message });
    }
  })
  .put("/:id", (req, res) => {
    const id = req.params.id;
    const token = req.headers.authorization.split(" ").at(-1);
    const myId = tokenUtils.extractId(token);
    const prescription = prescriptionService.getOne(id);
    if (myId !== prescription.doctorId) {
      return res.status(403).send({});
    }

    try {
      prescriptionService.updateOne(id, req.body);
      return res.status(200).send({});
    } catch (err) {
      return res.status(404).send({ message: err.message });
    }
  })
  .delete("/:id", (req, res) => {
    const id = req.params.id;
    const token = req.headers.authorization.split(" ").at(-1);
    const myId = tokenUtils.extractId(token);
    const prescription = prescriptionService.getOne(id);
    if (myId !== prescription.doctorId) {
      return res.status(403).send({});
    }

    try {
      prescriptionService.deleteOne(id);
      return res.status(200).send({});
    } catch (err) {
      return res.status(404).send({ message: err.message });
    }
  });

export default router;
